### ISSUE TITLE
Check type variable bounds of self types

### ISSUE THREAD

**Issue by JukkaL** [python/mypy#2351]
As follow-up to #2193, check that if self is declared with a type variable type, the type variable has a compatible bound. For example, some of these should be rejected:

```
T = TypeVar('T')
TA = TypeVar('TA', bound='A')
TC = TypeVar('TC', bound='C')

class A:
    def f(self: T) -> T: ...  # Should be error
    def g(self: TC) -> TC: ...  # Should be error
    def h(self: TA) -> TA: ...  # OK

class C: pass
```

(We may also want to check variance, but it's a separate issue and less clear.)

**Comment by elazarg** [comment]
#2381 should reject TC, but will accept T. I don't think I understand why should T be rejected.

**Comment by elazarg** [comment]
Another complication - if the function is overloaded, it is the union of the self-types that should be considered.

**Comment by elazarg** [comment]
@JukkaL do you consider the current behavior problematic/incomplete?

**Comment by JukkaL** [comment]
I don't think that allowing `T` is incorrect, but it's likely almost always not what the programmer intended, so giving a warning about it seems reasonable. Otherwise there could a more mysterious error messages generated elsewhere. This is not a very important issue, though.

**Comment by gvanrossum** [comment]
I've written code like this using just `T`. My assumption was that for that particular method, `T` would behave as if it had a bound of the containing class.

**Comment by elazarg** [comment]
Me too. I think the assumption about implicit bounds is "infer" rather than "object".

**Comment by JukkaL** [comment]
Hmm inferring the bound might be reasonable thing, at least when there is no explicit conflicting bound. However, that's not how self types work right now, and it would be a little ad hoc.
